Source/Controllers/Other/SPSQLiteHistoryManager.m

```diff
@@ -0,0 +1,437 @@
+		
+		if(error != nil){
+			SPLog(@"Problem opening db - %@ : %@",error.localizedDescription, error.localizedRecoverySuggestion);
+			return self; // ??
+		}
+		
+		SPLog(@"Is SQLite compiled with it's thread safe options turned on? %@!", [FMDatabase isSQLiteThreadSafe] ? @"Yes" : @"No");
+		
+		[self setupQueryHistoryDatabase];
+		
+		migratedPrefsToDB = user_defaults_get_bool_ud(SPMigratedQueriesFromPrefs, prefs);
+		if(migratedPrefsToDB == NO){
+			[self migrateQueriesFromPrefs]; // sets queryHist
+		}
+		else{
+			[self loadQueryHistory];
+		}
+		
+		[self getDBsize];
+		
+		SPLog(@"sqliteLibVersion = %@", [FMDatabase sqliteLibVersion]);
+		
+		return self;
+	}
+	
+	return nil;
+}
+
+- (long)idForRowAlreadyInDB:(NSString*)query{
+		
+	if(!queue){
+		SPLog(@"reopening queue");
+		queue = [FMDatabaseQueue databaseQueueWithPath:sqlitePath];
+	}
+	
+	long __block idForExistingRow = 0;
+	
+	[queue inDatabase:^(FMDatabase *db) {
+		
+		FMResultSet *rs = [db executeQuery:@"SELECT id FROM QueryHistory where query = ?", query];
+		while ([rs next]) {
+			idForExistingRow = [rs longForColumn:@"id"];
+		}
+		[rs close];
+	}];
+	
+	return idForExistingRow;
+}
+
+- (void)updateQueryHistory:(NSArray*)newHist{
+	
+	executeOnBackgroundThread(^{
+		
+		if(!self->queue){
+			SPLog(@"reopening queue");
+			self->queue = [FMDatabaseQueue databaseQueueWithPath:self->sqlitePath];
+		}
+		
+		BOOL __block success = NO;
+		
+		for(id obj in newHist){
+			
+			if([obj isKindOfClass:[NSString class]] && [(NSString *)obj length]){
+				// JCS - not sure we need this check .. just insert or ignore....
+				long idForExistingRow = [self idForRowAlreadyInDB:obj];
+				
+				if (idForExistingRow > 0){
+					
+					[self->queue inDatabase:^(FMDatabase *db) {
+						//db.traceExecution = YES;
+						
+						success = [db executeUpdate:@"UPDATE QueryHistory set modifiedTime = ? where id = ?", [NSDate date], @(idForExistingRow)];
+						
+						if (!success) {
+							FMDBQuickCheck([db hadError]);
+							if ([db hadError]) {
+								SPLog(@"Err %d: %@", [db lastErrorCode], [db lastErrorMessage]);
+							}
+						}
+//						SPLog(@"UPDATED = %ld, %@", idForExistingRow, obj);
+						// no else - nothing to update on the queryHist array
+					}];
+				}
+				else{
+					// if this is not unique then it's going to break
+					// we could check, but max 100 items ... probability of clash is low.
+					NSNumber *newKeyValue = [self primaryKeyValueForNewRow];
+					
+//					SPLog(@"newKeyValue: %@", newKeyValue);
+					
+					[self->queue inDatabase:^(FMDatabase *db) {
+						success = [db executeUpdate:@"INSERT OR IGNORE INTO QueryHistory (id, query, createdTime) VALUES (?, ?, ?)", newKeyValue, obj, [NSDate date]];
+						
+						if (success) {
+							[self->queryHist safeSetObject:obj forKey:newKeyValue];
+//							SPLog(@"INSERTED = %@, %@", newKeyValue, obj);
```

Comment:
some cleanups here?

Reply:
Concur here - the whole block about duplicates was the only part to me that looked like it may need some tweaks. Mostly I was confused about the uniqueness stuff?

Reply:
I figured a number would be quicker to check via the index rather than checking the long query string